File: platforms/javascript/src/core/execution/natives/NativeRegistry.js
// ©2026 JP Morgan Chase & Co. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Host natives for the JavaScript host, keyed by mangled native signature as on
// the JVM hosts. Most JavaScript natives are translation-time coders
// (translation.pure) and never pass through here; these are the ones
// runtime-lib hands to the host through a __host* hook. An implementation is a
// factory `(runtime) => function`: PureRuntime calls it once when it is built,
// so natives reach the registry, evalJs and the in-memory module without
// module-level state.

import { antlrExtension } from "./AntlrExtension.js";
import { compileSourceExtension } from "./CompileSourceExtension.js";
import { javascriptLanguageExtension } from "./JavaScriptLanguageExtension.js";
import {
    ANTLR_NATIVES, COMPILE_SOURCE, DIRECTORY_TREE, EVALUATE, JS_COMPILE_MODULE, JS_DRAIN_COMPILED_SOURCES, JS_EXECUTE, READ_FILE, READ_FILE_BYTES, WRITE_FILE,
} from "./native-signatures.js";

// The key each host native is installed under on the runtime's host object (globalThis.__pureHost).
const HOST_HOOKS = new Map([
    [COMPILE_SOURCE, "hostCompileSource"],
    [EVALUATE, "hostEvaluateFunctionDefinition"],
    [JS_COMPILE_MODULE, "hostJsCompile"],
    [JS_EXECUTE, "hostJsExecute"],
    [JS_DRAIN_COMPILED_SOURCES, "hostJsDrainCompiledSources"],
    [READ_FILE, "hostReadFile"],
    [READ_FILE_BYTES, "hostReadFileBytes"],
    [DIRECTORY_TREE, "hostDirectoryTree"],
    [WRITE_FILE, "hostWriteFile"],
    ...ANTLR_NATIVES.map(([name, signature]) => [signature, name]),
]);

/** The host-object key a native signature is installed under. */
export function hostHookFor(signature) {
    const hook = HOST_HOOKS.get(signature);
    if (!hook) throw new Error(`no runtime-lib hook for native ${signature}`);
    return hook;
}

// Built-ins every JavaScript host can run. Node-only extensions (FileSystemExtension) are added by
// the Node host (runtime/load-runtime.js), so browser pages can import this module.
export const DEFAULT_NATIVES_EXTENSIONS = [compileSourceExtension, javascriptLanguageExtension, antlrExtension];

export class NativeRegistry {
    #natives = new Map();

    /** The host's natives: every extension in `extensions` registers its own. */
    static createDefault(extensions = DEFAULT_NATIVES_EXTENSIONS) {
        const natives = new NativeRegistry();
        for (const extension of extensions) extension.registerAll(natives);
        return natives;
    }

    /** Register `impl`, a factory `(runtime) => function`, under a native signature. */
    register(signature, impl) {
        hostHookFor(signature);
        this.#natives.set(signature, impl);
        return this;
    }

    has(signature) { return this.#natives.has(signature); }
    get signatures() { return [...this.#natives.keys()]; }
    entries() { return this.#natives.entries(); }
}

File: platforms/truffle/legend-pure-next-truffle/legend-pure-next-truffle-core/src/main/java/org/finos/legend/pure/truffle/execution/natives/NativeRegistry.java
// Copyright 2024 Goldman Sachs
// ©2026 JP Morgan Chase & Co. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.finos.legend.pure.truffle.execution.natives;

import org.finos.legend.pure.truffle.execution.PureASTBuilder;

import org.finos.legend.pure.truffle.execution.ast.PureNode;

import java.util.HashMap;
import java.util.Map;
import java.util.ServiceLoader;

/**
 * Registry of specialized Truffle nodes, keyed by native signature (e.g.
 * {@code plus_Integer_MANY__Integer_1_}). {@link PureASTBuilder} resolves
 * every native call through this registry; an unregistered signature is an
 * error (a new native added without a corresponding Truffle node).
 *
 * <p>Specialized nodes live under
 * {@code org.finos.legend.pure.truffle.interpreter.ast.natives.**} and opt in by
 * registering a {@link Factory} here.</p>
 *
 * <p>A factory receives the already-lowered argument nodes plus the
 * {@code genericType} / {@code multiplicity} from the call site. The
 * default implementation is empty — Phase 2 registers math/boolean
 * natives, Phase 4 collection natives, etc.</p>
 */
public final class NativeRegistry
{
    /**
     * Factory for a specialized Truffle node covering one native signature.
     */
    @FunctionalInterface
    public interface Factory
    {
        PureNode create(PureNode[] args, Object genericType, Object multiplicity, Object fe);
    }

    private final Map<String, Factory> factories = new HashMap<>();

    /**
     * Default registry with no specializations.
     */
    public static NativeRegistry empty()
    {
        return new NativeRegistry();
    }

    /**
     * Creates a fully-populated registry with all standard native specializations.
     */
    public static NativeRegistry createDefault()
    {
        NativeRegistry registry = new NativeRegistry();
        org.finos.legend.pure.truffle.execution.natives.math.MathNodeFactories.registerAll(registry);
        org.finos.legend.pure.truffle.execution.natives.boolean_.BooleanNodeFactories.registerAll(registry);
        org.finos.legend.pure.truffle.execution.natives.string.StringNodeFactories.registerAll(registry);
        org.finos.legend.pure.truffle.execution.natives.collection.CollectionNodeFactories.registerAll(registry);
        org.finos.legend.pure.truffle.execution.natives.lang.LangNodeFactories.registerAll(registry);
        org.finos.legend.pure.truffle.execution.natives.meta.MetaNodeFactories.registerAll(registry);
        org.finos.legend.pure.truffle.execution.natives.meta.antlr.AntlrNodeFactories.registerAll(registry);
        org.finos.legend.pure.truffle.execution.natives.io.IONodeFactories.registerAll(registry);
        org.finos.legend.pure.truffle.execution.natives.pdb.BinaryNodeFactories.registerAll(registry);
        org.finos.legend.pure.truffle.execution.natives.lang.AssertNodeFactories.registerAll(registry);
        org.finos.legend.pure.truffle.execution.natives.date.DateNodeFactories.registerAll(registry);
        org.finos.legend.pure.truffle.execution.natives.variant.VariantNodeFactories.registerAll(registry);
        org.finos.legend.pure.truffle.execution.natives.javacompile.JavaCompileNodeFactories.registerAll(registry);
        // Extensions ship as separate jars (legend-pure-next-truffle-extension/*)
        // and contribute additional natives via the NativesExtension SPI.
        // Each jar's META-INF/services/org.finos.legend.pure.truffle.execution.natives.NativesExtension
        // names the implementation(s) ServiceLoader should construct here.
        for (NativesExtension ext : ServiceLoader.load(
                NativesExtension.class,
                NativeRegistry.class.getClassLoader()))
        {
            ext.registerAll(registry);
        }
        return registry;
    }

    public NativeRegistry register(String signature, Factory factory)
    {
        factories.put(signature, factory);
        return this;
    }

    /**
     * Returns the specialized factory for {@code signature}, or {@code null}
     * if the signature is still on the bridge.
     */
    public Factory lookup(String signature)
    {
        return factories.get(signature);
    }

    public boolean isEmpty()
    {
        return factories.isEmpty();
    }

    public int size()
    {
        return factories.size();
    }
}
